From d34b7de78259a2e83ec7cfc16bff084ad1d4685c Mon Sep 17 00:00:00 2001
From: Chaoste <t-kellermeier@hotmail.de>
Date: Thu, 4 Feb 2016 02:03:20 +0100
Subject: [PATCH 1/2] - Fix anchor links in profile (style + offset for navbar)
 - Add/Fix several back buttons (in edit views) - Remove workdays view for all
 projects - Add workdays view for all months in a project - Remove 'Show
 Timesheet' link in HiWi overview (missing function) - Only current_user can
 hand in timesheets - Only handed in timesheets can be seen by other users

---
 app/assets/stylesheets/application.css        | 14 +++++++++++
 app/controllers/work_days_controller.rb       | 33 ++++++++++++++-----------
 app/models/work_day.rb                        | 13 +++++-----
 app/views/chairs/show.html.erb                |  5 ----
 app/views/projects/show.html.erb              |  2 +-
 app/views/time_sheets/_show_footer.html.erb   | 26 +++++++++++---------
 app/views/time_sheets/edit.html.erb           |  9 ++++++-
 app/views/users/show.html.erb                 |  8 +++---
 app/views/work_days/edit.html.erb             |  2 +-
 app/views/work_days/index.html.erb            |  6 ++++-
 app/views/work_days/new.html.erb              |  2 +-
 config/locales/de.bootstrap.yml               |  1 +
 config/locales/en.bootstrap.yml               |  1 +
 spec/controllers/work_days_controller_spec.rb |  8 +-----
 spec/views/work_days/index.html.erb_spec.rb   | 35 ++++++++++++++-------------
 15 files changed, 95 insertions(+), 70 deletions(-)

diff --git a/app/assets/stylesheets/application.css b/app/assets/stylesheets/application.css
index fca8b96..bce05e6 100644
--- a/app/assets/stylesheets/application.css
+++ b/app/assets/stylesheets/application.css
@@ -67,6 +67,20 @@ body {
     margin-bottom: 15px;
 }
 
+/* Styles for anchors without a link (href) */
+a:not([href]):hover {
+    text-decoration: none;
+    background-color: rgba(255, 255, 255, 0);
+    color: black;
+}
+
+/* Add offset because the navbar covers the anchor element
+   (without adding a gap in the content) */
+a.anchor {
+    padding-top: 90px;
+    margin-top: -90px;
+}
+
 .projectCheckbox {
     margin-left: -30px;
 }
diff --git a/app/controllers/work_days_controller.rb b/app/controllers/work_days_controller.rb
index 6ace82e..70629c2 100644
--- a/app/controllers/work_days_controller.rb
+++ b/app/controllers/work_days_controller.rb
@@ -8,24 +8,33 @@ class WorkDaysController < ApplicationController
   end
 
   def index
-    if params.has_key?(:month) && params.has_key?(:year) && params.has_key?(:user_id) && User.find_by_id(params[:user_id]) != nil
-      @month = params[:month].to_i
-      @year = params[:year].to_i
-      @user = User.find(params[:user_id])
-      @project = params.has_key?(:project) && Project.find_by_id(params[:project].to_i) != nil ? Project.find(params[:project].to_i) : nil
-      @time_sheet = TimeSheet.time_sheet_for(@year, @month, @project, @user)
+    if params.has_key?(:project)
+      @month = params[:month].to_i # equals 0 when no month passed
+      @year = params[:year].to_i # equals 0 when no month passed
+      @user = params.has_key?(:user_id) && User.find_by_id(params[:user_id]) != nil ? User.find(params[:user_id]) : current_user
+      @project = Project.find_by_id(params[:project].to_i)
+      if @user.projects.find_by_id(@project.id) == nil
+        redirect_to user_path(current_user, anchor: 'timesheets')
+      end
+      if @year != 0 && @month != 0
+        @time_sheet = TimeSheet.time_sheet_for(@year, @month, @project, @user)
+        if current_user != @user && !@time_sheet.handed_in?
+          redirect_to user_path(@user)
+        end
+      end
       @work_days = WorkDay.all_for(@year, @month, @project, @user)
     else
-      date = Date.today
-      redirect_to work_days_path(month: date.month, year: date.year, user_id: current_user.id)
+      redirect_to user_path(current_user, anchor: 'timesheets')
     end
   end
 
   def show
+    redirect_to work_days_month_project_path
   end
 
   def new
     @work_day = WorkDay.new
+    @work_day.date = Date.today
   end
 
   def edit
@@ -56,7 +65,7 @@ def destroy
     date = @work_day.date
     @work_day.destroy
     flash[:success] = 'Work Day was successfully destroyed.'
-    redirect_to work_days_path(month: date.month, year: date.year)
+    redirect_to work_days_month_project_path
   end
 
   private
@@ -70,11 +79,7 @@ def work_day_params
     params.require(:work_day).permit(:date, :start_time, :break, :end_time, :duration, :attendance, :notes, :user_id, :project_id)
   end
 
-  def work_days_month_path
-    work_days_path(month: @work_day.date.month, year: @work_day.date.year)
-  end
-
   def work_days_month_project_path
-    work_days_path(month: @work_day.date.month, year: @work_day.date.year, project: @work_day.project.id, user_id: current_user.id)
+    work_days_path(month: @work_day.date.month, year: @work_day.date.year, project: @work_day.project.id, user_id: @work_day.user.id)
   end
 end
diff --git a/app/models/work_day.rb b/app/models/work_day.rb
index 3e35583..fa2b973 100644
--- a/app/models/work_day.rb
+++ b/app/models/work_day.rb
@@ -63,16 +63,17 @@ def project_id_exists
   end
 
   def self.all_for(year, month, project, user)
-    date = Date.new(year, month)
-    month_start = date.beginning_of_month
-    month_end = date.end_of_month
-    if project.nil?
-      return WorkDay.where(date: month_start..month_end, user: user)
+    # year and month are 0 when they are not passed to the controller (nil.to_i = 0)
+    if year == 0 || month == 0
+      return WorkDay.where(user: user, project_id: project)
     else
+      date = Date.new(year, month)
+      month_start = date.beginning_of_month
+      month_end = date.end_of_month
       return WorkDay.where(date: month_start..month_end, user: user, project_id: project)
     end
   end
-  
+
   def duration_hours_minutes
     work_time = duration_in_minutes
     minutes = work_time % 60
diff --git a/app/views/chairs/show.html.erb b/app/views/chairs/show.html.erb
index 29ec2d9..d82beca 100644
--- a/app/views/chairs/show.html.erb
+++ b/app/views/chairs/show.html.erb
@@ -68,11 +68,6 @@
         <% @chair.hiwis.each do |hiwi| %>
           <tr>
             <td><%= hiwi.name %></td>
-            <td>
-              <% if current_user.chair_wimi.representative %>
-                <%= I18n.t('chairs.navigation.show_timesheets') %>
-              <% end %>
-            </td>
           </tr>
         <% end %>
       </table>
diff --git a/app/views/projects/show.html.erb b/app/views/projects/show.html.erb
index 27d7106..247b1dc 100644
--- a/app/views/projects/show.html.erb
+++ b/app/views/projects/show.html.erb
@@ -193,7 +193,7 @@
             <p>
               <% if can? :edit, @project %>
                 <%= link_to t('projects.form.show_all_working_hours', default:'Show all working hours'),
-                              work_days_path(month: 1, year: 2016, project: @project)%>
+                              work_days_path(project: @project)%>
               <% end %>
             </p>
             <p>
diff --git a/app/views/time_sheets/_show_footer.html.erb b/app/views/time_sheets/_show_footer.html.erb
index d5c2e93..62990e3 100644
--- a/app/views/time_sheets/_show_footer.html.erb
+++ b/app/views/time_sheets/_show_footer.html.erb
@@ -123,20 +123,22 @@
           </div>
         </div>
         <div class="row vertical-align">
-          <div class="col-md-6">
-            <% unless @time_sheet.signed %>
-              <div class="control-group">
-                <%= f.check_box :signed , style: 'margin: 0px 10px;' %>
-                <%= f.label t('.add_signature') , style: 'margin: 0px;' %>
-              </div>
-            <% end %>
-          </div>
-          <div class="col-md-6">
-            <%= f.submit t('.hand_in'), class: 'btn btn-success btn-block' %>
-          </div>
+          <% if @time_sheet.user_id == current_user.id %>
+            <div class="col-md-6">
+              <% unless @time_sheet.signed %>
+                <div class="control-group">
+                  <%= f.check_box :signed , style: 'margin: 0px 10px;' %>
+                  <%= f.label t('.add_signature') , style: 'margin: 0px;' %>
+                </div>
+              <% end %>
+            </div>
+            <div class="col-md-6">
+              <%= f.submit t('.hand_in'), class: 'btn btn-success btn-block' %>
+            </div>
+          <% end %>
         </div>
       <% else %>
-        <% if @time_sheet.user_id = current_user.id %>
+        <% if @time_sheet.user_id == current_user.id %>
           <div class="row vertical-align">
             <div class="col-md-6">
               <% unless @time_sheet.signed %>
diff --git a/app/views/time_sheets/edit.html.erb b/app/views/time_sheets/edit.html.erb
index 5b012a1..f06c201 100644
--- a/app/views/time_sheets/edit.html.erb
+++ b/app/views/time_sheets/edit.html.erb
@@ -2,7 +2,14 @@
   <div class="display">
     <div class="page-header">
       <div class="h1">
-        <%= t('.title') %>
+        <div class="row">
+          <div class="col-md-10">
+            <%= t('.title') %>
+          </div>
+          <div class="col-md-2">
+            <%= link_to t('helpers.links.back'), work_days_path(month: @time_sheet.month, year: @time_sheet.year, project: @time_sheet.project_id, user_id: @time_sheet.user_id), class: "btn btn-default btn-xs"%>
+          </div>
+        </div>
       </div>
     </div>
     <%= render 'form' %>
diff --git a/app/views/users/show.html.erb b/app/views/users/show.html.erb
index 1237b54..afd3bda 100644
--- a/app/views/users/show.html.erb
+++ b/app/views/users/show.html.erb
@@ -94,7 +94,7 @@
       <div class="display">
         <div class="panel panel-default">
           <div class="panel-heading">
-            <h4><%= t('.time_sheets_overview')%></h4>
+            <h4><a class="anchor" name="timesheets"><%= t('.time_sheets_overview')%></a></h4>
           </div>
           <div class="panel-body">
             <div class="row">
@@ -102,7 +102,7 @@
                 <div class="col-md-6">
                   <div class="panel panel-default">
                     <div class="panel-heading">
-                      <%= link_to l(Date.new(month.first, month.last, 1), format: :without_day) , work_days_path(month: month.last, year: month.first) , class: "btn btn-info btn-block" %>
+                      <%= l(Date.new(month.first, month.last, 1), format: :without_day) %>:
                     </div>
                     <div class="panel-body">
                       <% current_user.projects_for_month(month.first, month.last).each do |project| %>
@@ -193,7 +193,7 @@
           <div class="panel-heading">
             <div class="row">
               <div class="col-sm-6">
-                <h4><a name="holidays"><%= t('.holiday_requests') %></a></h4>
+                <h4><a class="anchor" name="holidays"><%= t('.holiday_requests') %></a></h4>
               </div>
               <div class="col-sm-6 text text-right">
                 <% if @user == current_user %>
@@ -267,7 +267,7 @@
           <div class="panel-heading">
             <div class="row">
               <div class="col-md-6">
-                <h4><a name="trips"><%= t('.business_trips') %></a></h4>
+                <h4><a class="anchor" name="trips"><%= t('.business_trips') %></a></h4>
               </div>
               <div class="col-md-6 text text-right">
                 <% if @user == current_user %>
diff --git a/app/views/work_days/edit.html.erb b/app/views/work_days/edit.html.erb
index fa963ea..2d05256 100644
--- a/app/views/work_days/edit.html.erb
+++ b/app/views/work_days/edit.html.erb
@@ -8,7 +8,7 @@
             <%=t '.title', default: [:'helpers.titles.edit', 'Edit %{model}'], model: model_class.model_name.human.titleize %>
           </div>
           <div class="col-md-2">
-            <%= link_to t('helpers.links.back'), root_path , class: "btn btn-default btn-xs"%>
+            <%= link_to t('helpers.links.back'), work_days_path(month: @work_day.date.month, year: @work_day.date.year, project: @work_day.project.id, user_id: @work_day.user.id), class: "btn btn-default btn-xs"%>
           </div>
         </div>
       </h1>
diff --git a/app/views/work_days/index.html.erb b/app/views/work_days/index.html.erb
index f76ede9..bad46cd 100644
--- a/app/views/work_days/index.html.erb
+++ b/app/views/work_days/index.html.erb
@@ -7,7 +7,11 @@
       <h1>
         <div class="row">
           <div class="col-md-6">
-            <%= t('.working_hours_for') %> <%=l Date.new(@year, @month, 1), format: :without_day %>
+            <% if @year == 0 || @month == 0 %>
+              <%= t('.working_hours') %>
+            <% else %>
+              <%= t('.working_hours_for') %> <%=l Date.new(@year, @month, 1), format: :without_day %>
+            <% end %>
           </div>
           <div class="col-md-3 col-md-offset-3">
             <% unless @time_sheet.nil? %>
diff --git a/app/views/work_days/new.html.erb b/app/views/work_days/new.html.erb
index 450fa5f..0ac2757 100644
--- a/app/views/work_days/new.html.erb
+++ b/app/views/work_days/new.html.erb
@@ -8,7 +8,7 @@
             <%=t '.title', default: [:'helpers.titles.new', 'New %{model}'], model: model_class.model_name.human.titleize %>
           </div>
           <div class="col-md-2">
-            <%= link_to t('helpers.links.back'), root_path , class: "btn btn-default btn-xs"%>
+            <%= link_to t('helpers.links.back'), :back, class: "btn btn-default btn-xs"%>
           </div>
         </div>
       </h1>
diff --git a/config/locales/de.bootstrap.yml b/config/locales/de.bootstrap.yml
index e0838f4..73f95e7 100644
--- a/config/locales/de.bootstrap.yml
+++ b/config/locales/de.bootstrap.yml
@@ -286,6 +286,7 @@ de:
   work_days:
     index:
       working_hours_for: "Arbeitsstunden im"
+      working_hours: "Arbeitsstunden"
       new: "hinzufügen"
       for: "für"
       by: "von"
diff --git a/config/locales/en.bootstrap.yml b/config/locales/en.bootstrap.yml
index 0fb0108..d47ef83 100644
--- a/config/locales/en.bootstrap.yml
+++ b/config/locales/en.bootstrap.yml
@@ -296,6 +296,7 @@ en:
   work_days:
     index:
       working_hours_for: "Working hours for"
+      working_hours: "Working hours"
       new: "add workday"
       for: "for"
       by: "by"
diff --git a/spec/controllers/work_days_controller_spec.rb b/spec/controllers/work_days_controller_spec.rb
index b126864..9a30211 100644
--- a/spec/controllers/work_days_controller_spec.rb
+++ b/spec/controllers/work_days_controller_spec.rb
@@ -25,12 +25,6 @@
       expect(assigns(:work_days)).to eq([work_day])
     end
 
-    it 'redirects to work days of current month when no month is given' do
-      work_day = WorkDay.create! valid_attributes
-      get :index, {}, valid_session
-      expect(response).to redirect_to(work_days_path(month: Date.today.month, year: Date.today.year, user_id: @user.id))
-    end
-
     it 'shows work_days for month and project' do
       work_day = WorkDay.create! valid_attributes
       get :index, {month: Date.today.month, year: Date.today.year, user_id: @user.id, project_id: @project.id}, valid_session
@@ -147,7 +141,7 @@
     it 'redirects to the work_days list for current month' do
       work_day = WorkDay.create! valid_attributes
       delete :destroy, {id: work_day.to_param}, valid_session
-      expect(response).to redirect_to(work_days_path(month: work_day.date.month, year: work_day.date.year))
+      expect(response).to redirect_to(work_days_path(month: work_day.date.month, year: work_day.date.year, project: @project.id, user_id: @user.id))
     end
   end
 end
diff --git a/spec/views/work_days/index.html.erb_spec.rb b/spec/views/work_days/index.html.erb_spec.rb
index fc74779..34d8942 100644
--- a/spec/views/work_days/index.html.erb_spec.rb
+++ b/spec/views/work_days/index.html.erb_spec.rb
@@ -2,58 +2,59 @@
 
 RSpec.describe 'work_days/index.html.erb', type: :view do
   before :each do
-    @superadmin = FactoryGirl.create(:user, superadmin: true)
+    @user = FactoryGirl.create(:user)
     @chair = FactoryGirl.create(:chair)
-    ChairWimi.create(user: @superadmin, chair: @chair, representative: true)
-    login_as(@superadmin, scope: :user)
+    ChairWimi.create(user: @user, chair: @chair, representative: true)
+    login_as(@user, scope: :user)
     @project = FactoryGirl.create(:project)
+    @project.add_user(@user)
   end
 
   it 'expects a hand in button for not handed in timesheets' do
-    FactoryGirl.create(:time_sheet, user: @superadmin, project: @project, handed_in: false)
-    visit work_days_path(month: Date.today.month, year: Date.today.year, project: @project.id, user_id: @superadmin.id)
+    FactoryGirl.create(:time_sheet, user: @user, project: @project, handed_in: false)
+    visit work_days_path(month: Date.today.month, year: Date.today.year, project: @project.id, user_id: @user.id)
     expect(page).to have_selector("input[type=submit][value= 'hand in']")
   end
 
   it 'expects a accept button for handed in timesheets' do
-    FactoryGirl.create(:time_sheet, user: @superadmin, project: @project, handed_in: true)
-    visit work_days_path(month: Date.today.month, year: Date.today.year, project: @project.id, user_id: @superadmin.id)
+    FactoryGirl.create(:time_sheet, user: @user, project: @project, handed_in: true)
+    visit work_days_path(month: Date.today.month, year: Date.today.year, project: @project.id, user_id: @user.id)
     expect(page).to have_selector('input[type=submit][value= accept]')
   end
 
   it 'expects a reject button for handed in timesheets' do
-    FactoryGirl.create(:time_sheet, user_id: @superadmin.id, project_id: @project.id, handed_in: true)
-    visit work_days_path(month: Date.today.month, year: Date.today.year, project: @project.id, user_id: @superadmin.id)
+    FactoryGirl.create(:time_sheet, user_id: @user.id, project_id: @project.id, handed_in: true)
+    visit work_days_path(month: Date.today.month, year: Date.today.year, project: @project.id, user_id: @user.id)
     expect(page).to have_selector('input[type=submit][value= reject]')
   end
 
   it 'rejects a TimeSheet if reject button is pressed' do
-    timesheet = FactoryGirl.create(:time_sheet, user: @superadmin, project: @project, handed_in: true)
-    visit work_days_path(month: Date.today.month, year: Date.today.year, project: @project.id, user_id: @superadmin.id)
+    timesheet = FactoryGirl.create(:time_sheet, user: @user, project: @project, handed_in: true)
+    visit work_days_path(month: Date.today.month, year: Date.today.year, project: @project.id, user_id: @user.id)
     click_on('reject')
     timesheet.reload
     expect(timesheet.status).eql? 'rejected'
   end
 
   it 'accepts a TimeSheet if accept button is pressed' do
-    timesheet = FactoryGirl.create(:time_sheet, user: @superadmin, project: @project, handed_in: true)
-    visit work_days_path(month: Date.today.month, year: Date.today.year, project: @project.id, user_id: @superadmin.id)
+    timesheet = FactoryGirl.create(:time_sheet, user: @user, project: @project, handed_in: true)
+    visit work_days_path(month: Date.today.month, year: Date.today.year, project: @project.id, user_id: @user.id)
     click_on('accept')
     timesheet.reload
     expect(timesheet.status).eql? 'accepted'
   end
 
   it 'hands in a TimeSheet if hand in button is pressed' do
-    timesheet = FactoryGirl.create(:time_sheet, user: @superadmin, project: @project, handed_in: false)
-    visit work_days_path(month: Date.today.month, year: Date.today.year, project: @project.id, user_id: @superadmin.id)
+    timesheet = FactoryGirl.create(:time_sheet, user: @user, project: @project, handed_in: false)
+    visit work_days_path(month: Date.today.month, year: Date.today.year, project: @project.id, user_id: @user.id)
     click_on('hand in')
     timesheet.reload
     expect(timesheet.handed_in).to be true
   end
 
   it 're hands in a TimeSheet if sign button is pressed' do
-    timesheet = FactoryGirl.create(:time_sheet, user: @superadmin, project: @project, handed_in: false, status: 'rejected', signer: @superadmin.id)
-    visit work_days_path(month: Date.today.month, year: Date.today.year, project: @project.id, user_id: @superadmin.id)
+    timesheet = FactoryGirl.create(:time_sheet, user: @user, project: @project, handed_in: false, status: 'rejected', signer: @user.id)
+    visit work_days_path(month: Date.today.month, year: Date.today.year, project: @project.id, user_id: @user.id)
     click_on('hand in')
     timesheet.reload
     expect(timesheet.status).eql? ('rejected')

From 42f5510f8524b0d18909cbdebd6546b4e6d30cbb Mon Sep 17 00:00:00 2001
From: Theresa <theresa.zobel@student.hpi.de>
Date: Thu, 4 Feb 2016 10:47:34 +0100
Subject: [PATCH 2/2] some refactoring

---
 app/views/time_sheets/_show_footer.html.erb | 4 ++--
 app/views/time_sheets/edit.html.erb         | 2 +-
 app/views/work_days/edit.html.erb           | 2 +-
 app/views/work_days/new.html.erb            | 2 +-
 4 files changed, 5 insertions(+), 5 deletions(-)

diff --git a/app/views/time_sheets/_show_footer.html.erb b/app/views/time_sheets/_show_footer.html.erb
index 62990e3..71ff02c 100644
--- a/app/views/time_sheets/_show_footer.html.erb
+++ b/app/views/time_sheets/_show_footer.html.erb
@@ -123,7 +123,7 @@
           </div>
         </div>
         <div class="row vertical-align">
-          <% if @time_sheet.user_id == current_user.id %>
+          <% if @time_sheet.user == current_user %>
             <div class="col-md-6">
               <% unless @time_sheet.signed %>
                 <div class="control-group">
@@ -138,7 +138,7 @@
           <% end %>
         </div>
       <% else %>
-        <% if @time_sheet.user_id == current_user.id %>
+        <% if @time_sheet.user == current_user %>
           <div class="row vertical-align">
             <div class="col-md-6">
               <% unless @time_sheet.signed %>
diff --git a/app/views/time_sheets/edit.html.erb b/app/views/time_sheets/edit.html.erb
index f06c201..9b41892 100644
--- a/app/views/time_sheets/edit.html.erb
+++ b/app/views/time_sheets/edit.html.erb
@@ -7,7 +7,7 @@
             <%= t('.title') %>
           </div>
           <div class="col-md-2">
-            <%= link_to t('helpers.links.back'), work_days_path(month: @time_sheet.month, year: @time_sheet.year, project: @time_sheet.project_id, user_id: @time_sheet.user_id), class: "btn btn-default btn-xs"%>
+            <%= link_to t('helpers.links.back'), work_days_path(month: @time_sheet.month, year: @time_sheet.year, project: @time_sheet.project_id, user_id: @time_sheet.user_id), class: 'btn btn-default btn-xs'%>
           </div>
         </div>
       </div>
diff --git a/app/views/work_days/edit.html.erb b/app/views/work_days/edit.html.erb
index 2d05256..5563c6a 100644
--- a/app/views/work_days/edit.html.erb
+++ b/app/views/work_days/edit.html.erb
@@ -8,7 +8,7 @@
             <%=t '.title', default: [:'helpers.titles.edit', 'Edit %{model}'], model: model_class.model_name.human.titleize %>
           </div>
           <div class="col-md-2">
-            <%= link_to t('helpers.links.back'), work_days_path(month: @work_day.date.month, year: @work_day.date.year, project: @work_day.project.id, user_id: @work_day.user.id), class: "btn btn-default btn-xs"%>
+            <%= link_to t('helpers.links.back'), work_days_path(month: @work_day.date.month, year: @work_day.date.year, project: @work_day.project.id, user_id: @work_day.user.id), class: 'btn btn-default btn-xs'%>
           </div>
         </div>
       </h1>
diff --git a/app/views/work_days/new.html.erb b/app/views/work_days/new.html.erb
index 0ac2757..1b48066 100644
--- a/app/views/work_days/new.html.erb
+++ b/app/views/work_days/new.html.erb
@@ -8,7 +8,7 @@
             <%=t '.title', default: [:'helpers.titles.new', 'New %{model}'], model: model_class.model_name.human.titleize %>
           </div>
           <div class="col-md-2">
-            <%= link_to t('helpers.links.back'), :back, class: "btn btn-default btn-xs"%>
+            <%= link_to t('helpers.links.back'), :back, class: 'btn btn-default btn-xs'%>
           </div>
         </div>
       </h1>
